core/src/main/scala/org/apache/spark/sql/rapids/tool/qualification/QualificationAppInfo.scala

```diff
@@ -200,22 +200,79 @@ class QualificationAppInfo(
 
   private def getStageToExec(execInfos: Seq[ExecInfo]): (Map[Int, Seq[ExecInfo]], Seq[ExecInfo]) = {
     val execsWithoutStages = new ArrayBuffer[ExecInfo]()
-    val perStageSum = execInfos.flatMap { execInfo =>
-      if (execInfo.stages.size > 1) {
-        execInfo.stages.map((_, execInfo))
-      } else if (execInfo.stages.size < 1) {
-        // we don't know what stage its in or its duration
-        logDebug(s"No stage associated with ${execInfo.exec} " +
-          s"so speedup factor isn't applied anywhere.")
-        execsWithoutStages += execInfo
-        Seq.empty
-      } else {
-        Seq((execInfo.stages.head, execInfo))
-      }
-    }.groupBy(_._1).map { case (stage, execInfos) =>
+
+    // This is to get the mapping between stageId and execs. This is primarily done based on
+    // accumulatorId. If an Exec has some metrics generated, then an accumulatorId will be
+    // generated for that Exec. This accumulatorId is used to get the stageId. If an Exec
+    // doesn't have any metrics, then we will try to get the stageId by looking at the
+    // neighbor Execs. If either of the neighbor Execs has a stageId, then we will use that
+    // to assign the same stageId to the current Exec as it's most likely that the current
+    // Exec is part of the same stage as the neighbor Exec.
+    val execInfosInOrder = execInfos.reverse
+    val execInfosCombined = Iterator(Seq(execInfosInOrder.head)) ++
+      execInfosInOrder.sliding(3) ++ Iterator(Seq(execInfosInOrder.last))
+    val perStageSum = execInfosCombined.map {
```

Comment:
Can this change cause double counting? 

I was testing on `Q23_part1`, stagesIds `1420` and `1400` both show 9 Execs and `AQEShuffleRead` is present in both stages. However in the DAG, `AQEShuffleRead` is present only in stage `1420` and not `1400`.


![Pasted Graphic 1](https://github.com/NVIDIA/spark-rapids-tools/assets/13639815/49c47db6-5a5a-4d52-8ae5-939723288ef2)

Reply:
Thanks for testing. It was double counting around Exchange nodes. Updated it.